### PR TITLE
docs/major update to docs, sphinx, and docs ci script

### PR DESCRIPTION
**Pull request by madeline-scyphers**
- [Update ModularMetric to directly save metric_to_eval instead of using…](https://github.com/madeline-scyphers/boa/commit/98b1e908aede600d4360cb4fdc92b320ab535645)  …
  … additional class
  
  ModularMetric use to use a MetricToEval class for json serialization, which wasn't necessary
  at all. Removed that class, and moved the functionality onto ModularMetric. This reduced
  clutter and confusion. Now when you look at a metric, metric.f/metric.metric_to_eval is
  the function, instead of having todo metric.f.func/metric.metric_to_eval.func Which feels
  much easier and cleaner. This also made it easier to make all the built--in metrics
  subclasses, which allowed easier documentation with sphinx rendering.
  
  Added a test that saves and loads a scheduler and resumes an experiment to check that
  the scheduler saving works. Added another test to save an arbitrary callable that is
  passed in. it doesn't work right now, marked as skip and made an issue to get that
  functionality for scheduler saving and loading with arbitrary callables working at
- [Modify api template](https://github.com/madeline-scyphers/boa/commit/11ee4f3b7595c8ee746347c500a1912b5b652601)  …
  Modified code reference api section to not use the
  template as much, and allowing each module to be generated
  when it is needed in the api section. This is a tradeoff of
  convencience for more customizability and a better doc design
- [More docstring updates, fix broken rst links](https://github.com/madeline-scyphers/boa/commit/8cdcef82d3d8325fdc0f2d528d86f4257f12d4e6)  …
  Add docutils func that adds a reference to the relative
  __init__.py of all files __doc__ in each submodule
  
  change sphinx configuration to be nitpicky which revealed a bunch
  of broken sphinx links which were then fixed (except ax and
  botorch links which can't be fixed right now because
  they don't build sphinx in a normal way.
  see https://github.com/facebook/Ax/issues/1227
  and see https://github.com/pytorch/botorch/issues/1464
  for more info on when ax and botroch can be linked).
- [Add examples, boa cli info, config info](https://github.com/madeline-scyphers/boa/commit/d74f7bdc1d017687172476a55bd7cc4cec4d26ec)
  Add example cli boa script with output and basic instructions on how to do boa cli
- [Add example notebooks and add docs build to CI](https://github.com/madeline-scyphers/boa/pull/36/commits/93444ad91204683a9703f76ae70190e92f8c9315) 
  Add an example notebook that runs a toy example in python and
  saves the output and another one that loads that output
  from the scheduler.json file and then displays the plots

### CODE REVIEW

File: boa/controller.py [hunk 1]
```diff
@@ -1,37 +1,91 @@
...
+        Sets up all the classes and objects needed to create the ax Scheduler
+
+        Parameters
+        ----------
+        append_timestamp
+            whether to append the output experiment directory with a timestamp or not
+            (default True)
+        experiment_dir
+            output experiment directory to save the experiment and trials to
+            (defaults to what is specified in the config,
```
Comment (jemissik) [inline boa/controller.py:65]: Probably we need to add some examples/instructions about these options to the configuration guide

File: boa/metrics/__init__.py [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
 
-    Coming soon!
+
+Metrics are the objective functions we are optimizing over
+or functions that we are adding in as further constraints of our problem.
+
+Metrics can be specified in your configuration by detailing the name
+of one of the predefined :mod:`Metrics <boa.metrics.metrics>` in BOA
+by passing in the name of the metric into your configuration file
+
```
Comment (jemissik) [inline boa/metrics/__init__.py:13]: We should add a note here that points to where users need to look if they are using custom metrics.

Comment (jemissik) [reply boa/metrics/__init__.py:13]: And also add mention of this to the user guide. I don't see anywhere obvious that points out that you can pass your own metric through directly if you need more customization

File: docs/user_guide/getting_started.rst [hunk 3]
```diff
@@ -28,6 +29,35 @@ installed. There are two options for this:
...
+
+If you want to install the latest (bleeding-edge) develop version of boa::
+
+    pip install git+https://github.com/madeline-scyphers/boa.git@develop
+
+********************************
+Installing for Contributing
+********************************
+
 Clone the boa repository from `boa's GitHub page <https://github.com/madeline-scyphers/boa>`_.
```
Comment (jemissik) [inline docs/user_guide/getting_started.rst:61]: Maybe should make a note to fork the repo, if it's for contributors?

**Review by jemissik — CHANGES_REQUESTED** [review]
I pushed a few edits to the docs, and fixed the refs in the notebooks

File: docs/user_guide/getting_started.rst [hunk 4]
```diff
@@ -28,7 +29,37 @@ installed. There are two options for this:
...
+If you don't need to create a new environment, activate the existing conda environment you will be using.
+
+If you are not on an x86 mac (or a mac with python running through rosetta), run these commands to install the dependencies::
+
+    conda install botorch -c pytorch -c gpytorch -c conda-forge
+    pip install ax-platform
+
+x86 macs (or a mac with python running through rosetta), run::
+
+    conda install botorch -c pytorch-nightly -c pytorch -c gpytorch -c conda-forge
```
Comment (jemissik) [inline docs/user_guide/getting_started.rst:46]: Needs updating for x86?

Comment (madeline-scyphers) [reply docs/user_guide/getting_started.rst]: Good catch

**Review by jemissik — APPROVED** [review]

**Pull request merged**